python/tvm/octo/inject_op_pattern.py

```diff
@@ -0,0 +1,84 @@
+# Licensed to the Apache Software Foundation (ASF) under one
+# or more contributor license agreements.  See the NOTICE file
+# distributed with this work for additional information
+# regarding copyright ownership.  The ASF licenses this file
+# to you under the Apache License, Version 2.0 (the
+# "License"); you may not use this file except in compliance
+# with the License.  You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing,
+# software distributed under the License is distributed on an
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+# pylint: disable=invalid-name, abstract-method, unused-argument
+"Operator based rewriting of fusion pattern."
+import re
+import enum
+import tvm
+from tvm import relax
+from tvm.relax.expr_functor import PyExprMutator
+
+
+class OpPatternKind(enum.IntEnum):
+    """Helpful naming for each fusion pattern"""
+
+    kElemWise = 0
+    kBroadcast = 1
+    kInjective = 2
+    kCommReduce = 3
+    kOutEWiseFusable = 4
+    kTuple = 7
+    kOpaque = 8
+
+
+@tvm.ir.transform.module_pass(opt_level=0)
+class InjectOpPattern(tvm.transform.Pass):
+    """Transformation pass that changes the fusion pattern for certain ops."""
+
+    def transform_module(self, mod: tvm.IRModule, ctx: tvm.transform.PassContext) -> tvm.IRModule:
+        """
+        Iterate through a module and explicitly adjust the fusion pattern for some ops.
+        """
+
+        @relax.expr_functor.visitor
+        class AdjustOpPattern(PyExprMutator):
+            """Visitor that changes the fusion pattern of specified operators."""
+
+            def __init__(self, mod):
+                self.mod = mod
+                # Create a map between functions and the desired pattern to register.
+                self.pattern_map = {
+                    "layer_norm": OpPatternKind.kInjective,
+                    "take": OpPatternKind.kInjective,
+                    "split": OpPatternKind.kInjective,
```

Comment:
This looks good for now, but should we consider injecting `op_pattern` in LegalizeOps pass? The way we will do this is keep `op_pattern` as optional property on Relax ops and we inject `op_pattern` during lowering to the PrimFunc representation of the op.

Reply:
+1. Agreed. Legalization seems to be the right place. With this future change, I guess we would be able to remove this pass as well as `AnnotateTIROpPattern()`?

Also, it seems like these are corner cases for `AnnotateTIROpPattern()`. Let's revisit this to fix when we have a bandwidth.

Reply:
Yeah this is a short term fix for things AnnotateTIROpPattern is missing.